coding.go

```diff
@@ -34,64 +34,63 @@ func (nb _PBNode__NodeBuilder) DecodeDagProto(r io.Reader) error {
 	if err := pbn.Unmarshal(encoded); err != nil {
 		return fmt.Errorf("unmarshal failed. %v", err)
 	}
-	pbLinks := make([]PBLink, 0, len(pbn.Links))
-	for _, link := range pbn.Links {
-		hash, err := cid.Cast(link.GetHash())
+	return fluent.Recover(func() {
+		fb := fluent.WrapAssembler(nb)
+		fb.CreateMap(0, func(fmb fluent.MapAssembler) {
+			fmb.AssembleEntry("Links").CreateList(len(pbn.Links), func(flb fluent.ListAssembler) {
+				for _, link := range pbn.Links {
+					hash, err := cid.Cast(link.GetHash())
 
-		if err != nil {
-			return fmt.Errorf("unmarshal failed. %v", err)
-		}
-		pbLinks = append(pbLinks, PBLink{
-			d: PBLink__Content{
-				Hash: MaybeLink{
-					Maybe: schema.Maybe_Value,
-					Value: Link{cidlink.Link{Cid: hash}},
-				},
-				Name: MaybeString{
-					Maybe: schema.Maybe_Value,
-					Value: String{link.GetName()},
-				},
-				Tsize: MaybeInt{
-					Maybe: schema.Maybe_Value,
-					Value: Int{int(link.GetTsize())},
-				},
-			},
+					if err != nil {
+						panic(fluent.Error{Err: fmt.Errorf("unmarshal failed. %v", err)})
```

Comment:
@rvagg : I went deep on exactly what happens here if the protobuf has a nil value for hash, because of its relevance to our other discussions about documenting this behavior:

tl;dr: a lack of hash here is not actually tolerated by this code: we get an error returned here.

(In detail: `cid.Cast` passes the nil down until some varint code tries to read from it, and at that point we get `cid.ErrVarintBuffSmall` and that bubbles back up from here.)

So I think this is a pretty solid argument that we *don't* have to consider `Hash` an `optional` in our IPLD Schema describing this.